src/codegen/encoding/mod.rs

```diff
@@ -546,6 +607,567 @@ pub(super) trait AbiEncoding {
+    /// containing the read value and the number of bytes read.
+    fn read_from_buffer(
+        &self,
+        buffer: &Expression,
+        offset: &Expression,
+        ty: &Type,
+        validator: &mut BufferValidator,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> (Expression, Expression) {
+        match ty {
+            Type::Uint(width) | Type::Int(width) => {
+                let encoding_size = width.next_power_of_two();
+
+                let size = Expression::NumberLiteral(Codegen, Uint(32), (encoding_size / 8).into());
+                validator.validate_offset_plus_size(offset, &size, ns, vartab, cfg);
+
+                let read_value = Expression::Builtin(
+                    Codegen,
+                    vec![ty.clone()],
+                    Builtin::ReadFromBuffer,
+                    vec![buffer.clone(), offset.clone()],
+                );
+                let read_var = vartab.temp_anonymous(ty);
+
+                cfg.add(
+                    vartab,
+                    Instr::Set {
+                        loc: Codegen,
+                        res: read_var,
+                        expr: if encoding_size == *width {
+                            read_value
+                        } else {
+                            Expression::Trunc(Codegen, ty.clone(), Box::new(read_value))
+                        },
+                    },
+                );
+
+                let read_expr = Expression::Variable(Codegen, ty.clone(), read_var);
+                (read_expr, size)
+            }
+
+            Type::Bool
+            | Type::Address(_)
+            | Type::Contract(_)
+            | Type::Enum(_)
+            | Type::Value
+            | Type::Bytes(_) => {
+                let read_bytes = ty.memory_size_of(ns);
+
+                let size = Expression::NumberLiteral(Codegen, Uint(32), read_bytes);
+                validator.validate_offset_plus_size(offset, &size, ns, vartab, cfg);
+
+                let read_value = Expression::Builtin(
+                    Codegen,
+                    vec![ty.clone()],
+                    Builtin::ReadFromBuffer,
+                    vec![buffer.clone(), offset.clone()],
+                );
+
+                let read_var = vartab.temp_anonymous(ty);
+                cfg.add(
+                    vartab,
+                    Instr::Set {
+                        loc: Codegen,
+                        res: read_var,
+                        expr: read_value,
+                    },
+                );
+
+                let read_expr = Expression::Variable(Codegen, ty.clone(), read_var);
+
+                (read_expr, size)
+            }
+
+            Type::DynamicBytes | Type::String => {
+                // String and Dynamic bytes are encoded as size + elements
+                let (array_length_var, size_length) =
+                    self.retrieve_array_length(buffer, offset, vartab, cfg);
```

Comment:
LLVM does not differentiate between signed and unsigned integers. So, let's say that we don't have the last four bytes to read when retrieving the length, but only two (offset + 2). We can read a -3 (remember, in LLVM everything is signed), so we would validate an offset of 4 (size width) + (-3) + offset = offset + 1. This is within bounds. Now we have an access violation when trying to read the string.

Check the example below. Imagine the buffer has only two bytes (an int16), but you are going to read four bytes, because no validation happens. If the extra two bytes you read happen to be `[255, 255]`, the number you read is -4. Your validator will try to validate 4 + (-4) + offset = offset, which is within bounds. When we try to read the string, we are going to have a segmentation fault (Access violation on Solana VM).

```Rust
#[test]
fn validation() {
    #[derive(Debug, BorshSerialize)]
    struct Input {
        item_1: i16,
        item_2: i16, // Rubbish: imagine this as the extra two bytes you are reading.
    }

    let mut vm = build_solidity(
        r#"
    contract Testing {
        function test(bytes memory buffer) public view returns (string) {
            string ss = abi.decode(buffer, string);

            return ss;
        }
    }
        "#,
    );

    vm.constructor(&[]);

    let input = Input {
        item_1: -4,
        item_2: -1,
    };
    let encoded = input.try_to_vec().unwrap();

    let res = vm.function("test", &[BorshToken::Bytes(encoded)]);
    std::println!("{:?}\n", res);
}

```

Reply:
> 2. I was told that the buffer always has some margin at the end anyways.

I think there might be space between heap allocated memory, but we cannot always assume that. The allocate functions we have allocate the exact number of bytes we request.

Reply:
> > 2. I was told that the buffer always has some margin at the end anyways.
> 
> I think there might be space between heap allocated memory, but we cannot always assume that. The allocate functions we have allocate the exact number of bytes we request.

The current heap implementation does always have some bytes at the end, for house keeping.

Reply:
> > LLVM does not differentiate between signed and unsigned integers
> 
> Are numbers in LLVM always signed? I don't get it, surely you can do unsigned operations? Why are we using a signed number for the offset then, while it doesn't make sense for something like a buffer offset to be negative first place?

That's not strictly true. Variables / values are neither signed nor unsigned. Operations are signed or unsigned like signed div/unsigned div. Add/Sub/Mul are not signed either, but the overflow checking versions are

Reply:
Thanks, it is how I expect this to work then, and so the scale version then should also work even for bogus data (given #1216 first):  
- Parsing a smaller value should trap the contract latest when the decoding is finished, because not all bytes were read
- Parsing a larger value should trap the contract because of the following OOB read

@LucasSte WDYT?

Reply:
I think that's right Cyril. We still need to validate if we haven't read rubbish data. After reading the `array_length`, we should check if `offset + size_width` (please, don't call this `size_length`) is within bounds. This check should happen before the validation of `offset + size_width + array_length`.